### PR TITLE
Add blog post: LibrePCB has switched to S-Expressions

### PR DESCRIPTION
**Pull request by ubruhin**
[View rendered blog post](https://github.com/LibrePCB/LibrePCB.github.io/blob/blogpost/librepcb_has_switched_to_s_expressions/content/blog/2017-10-30_librepcb_has_switched_to_s_expressions.md)

Should be published only after https://github.com/LibrePCB/LibrePCB/pull/193 is merged ;)

### CODE REVIEW

File: content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+---
+date: 2017-10-29
+title: LibrePCB has switched to S-Expressions
+author: U. Bruhin
+---
+
+One major goal of LibrePCB is to use human readable files to represent libraries,
+projects and so on. This is important to let the user understand what the EDA
+tool actually does when modifying a schematic or board. In addition, a good file
+format is essential to use version control systems to track changes in libraries
```
Comment (dbrgn) [inline content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md]: What about `s/to use/to be able to use/`?

File: content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+### XML and similar formats
+
+Until now, LibrePCB has used [XML](https://en.wikipedia.org/wiki/XML) as file
+format. But unfortunately we were not very happy with it, because of several
+reasons:
+
+- The XML specifications are **rather complex** (we used only a small subset of
+  the available features).
+- The (ugly) syntax is **quite verbose** (contains even redundant information,
```
Comment (dbrgn) [inline content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md]: contains even -> it even contains

File: content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md (hunk 1 above)
Comment (ubruhin) [reply content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md]: Or simply `s/to use/for/`?

**Comment by ubruhin** [comment]
> I'd also mention how much easier to parse S expressions are, compared to XML.

I'm not sure about that. In my opinion the complexity of a parser is irrelevant, as long as it is only inside an existing, properly working library, so we don't have to worry about it. And from our point of view, it doesn't really make a difference whether we use an XML parser or an S-Expressions parser, both is quite simple to use.

Instead, I prefer to focus on the look/formatting of the files we generate, because this is *much* more important (especially for *users* of LibrePCB) than the complexity of the parser.

> It might also be fair to give some credit to KiCad, as we were at least in part inspired by their format choice.

Good point, I'll try to add this :+1:

File: content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
+---
+date: 2017-10-29
```
Comment (rnestler) [inline content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md]: Maybe use the current date?

File: content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+title: LibrePCB has switched to S-Expressions
+author: U. Bruhin
+---
+
+One major goal of LibrePCB is to use human readable files to represent libraries,
+projects and so on. This is important to let the user understand what the EDA
+tool actually does when modifying a schematic or board. In addition, a good file
+format is essential for version control systems to track changes in libraries
```
Comment (rnestler) [inline content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md]: Maybe change to "to keep track of changes"?

File: content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+and projects.
+
+But LibrePCB doesn't just want to have *some* human readable file format. It
+wants to use the **best possible** file format for our specific needs!
+
+### XML and similar formats
+
+Until now, LibrePCB has used [XML](https://en.wikipedia.org/wiki/XML) as file
```
Comment (rnestler) [inline content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md]: /s/has used/used/

File: content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+format. But unfortunately we were not very happy with it, because of several
+reasons:
+
+- The XML specifications are **rather complex** (we used only a small subset of
+  the available features). In [this blog post](https://blog.pragmatists.com/xml-be-cautious-69a981fdc56a)
+  you can see that this complexity can even be dangerous!
+- The (ugly) syntax is **quite verbose** (it even contains redundant information,
```
Comment (rnestler) [inline content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md]: I'd remove the "(ugly)"

File: content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+the only reasonable solution is to use a custom file format, or at least a
+custom generator.
+
+### S-Expressions with custom generator
+
+To avoid completely reinventing the wheel, we decided to use the
+[S-Expressions](http://people.csail.mit.edu/rivest/Sexp.txt) file format.
+[KiCad](http://kicad-pcb.org) uses this file format
+[since version 4.0](http://kicad-pcb.org/post/release-4.0.0/), and it seems that it's
+really good suited for the use case of an EDA tool.
```
Comment (rnestler) [inline content/blog/2017-10-29_librepcb_has_switched_to_s_expressions.md]: /s/good/well/

File: content/blog/2017-10-30_librepcb_has_switched_to_s_expressions.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
+---
+date: 2017-10-30
+title: LibrePCB has switched to S-Expressions
```
Comment (rnestler) [inline content/blog/2017-10-30_librepcb_has_switched_to_s_expressions.md]: Maybe use [Title Case](http://www.grammar-monster.com/lessons/capital_letters_title_case.htm)?

File: content/blog/2017-10-30_librepcb_has_switched_to_s_expressions.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+One major goal of LibrePCB is to use human readable files to represent libraries,
+projects and so on. This is important to let the user understand what the EDA
+tool actually does when modifying a schematic or board. In addition, a good file
+format is essential for version control systems to keep track of changes in
+libraries and projects.
+
+But LibrePCB doesn't just want to have *some* human readable file format. It
+wants to use the **best possible** file format for our specific needs!
+
+### XML and similar formats
```
Comment (rnestler) [inline content/blog/2017-10-30_librepcb_has_switched_to_s_expressions.md]: Also maybe Title Case?

File: content/blog/2017-10-30_librepcb_has_switched_to_s_expressions.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+- Available XML generators typically provide **only little control over the
+  formatting** of the generated XML files. This leads to either very long files
+  (many linebreaks) or very long lines (only few linebreaks). Both is bad for
+  readability.
+
+Other file formats (or their generators) have similar issues, so it seems that
+the only reasonable solution is to use a custom file format, or at least a
+custom generator.
+
+### S-Expressions with custom generator
```
Comment (rnestler) [inline content/blog/2017-10-30_librepcb_has_switched_to_s_expressions.md]: Also maybe Title Case?

**Comment by ubruhin** [comment]
Thanks for your reviews!

**Pull request merged**